### PR TITLE
Refactor main_tf2

### PR DESCRIPTION
**Pull request by kartik4949**
Refactored main_tf2
def main function into a trainer class with helper functions for better readability.

### CODE REVIEW

**Comment by kartik4949** [comment]
In-Progress

**Comment by kartik4949** [comment]
need to add main_tf2 test

File: efficientnetv2/main_tf2.py [hunk 1]
```diff
@@ -118,189 +92,321 @@ def test_step(self, data):
...
-        logging.info('Epoch: %d, total %d', current_epoch, config.train.epochs)
-        if current_epoch >= config.train.epochs:
-          break
+
+  def evaluate(self) -> None:
+    """ Evaluation helper function. """
+    for ckpt in tf.train.checkpoints_iterator(
+        FLAGS.model_dir, timeout=60 * 60 * 24):
+      model.load_weights(ckpt)
+      eval_results = model.evaluate(
```
Comment (kartik4949) [inline efficientnetv2/main_tf2.py:294]: @mingxingtan eval_results we are not printing?

Comment (mingxingtan) [reply efficientnetv2/main_tf2.py:322]: Oh, you can print it out.

File: efficientnetv2/main_tf2.py [hunk 2]
```diff
@@ -118,189 +92,349 @@ def test_step(self, data):
...
+    return ds_strategy
+
+
+
+
+
+
+
+
+
```
Comment (mingxingtan) [inline efficientnetv2/main_tf2.py:437]: Too many empty lines?

Comment (kartik4949) [reply efficientnetv2/main_tf2.py:437]: Sorry will remove em 
PR is still in progress

**Comment by kartik4949** [comment]
writing smoke tests for main_tf2

**Comment by LucasSloan** [comment]
Thanks for sending this cl over!

I don't think this is the direction we want to go with the tf2 training script - it's written to mirror, as much as possible, the tf1 training script, so drastically rearchitecting it to be class makes the similarities harder to see.

If you could pull out common functions between the 2 training scripts and move them to utils, I think that could work better.

**Comment by kartik4949** [comment]
well, we can make tf1 and tf2 both in a better design pattern aesthetic :) 
but if that's too much work then I can just move on and find similar functions as u said.

File: efficientnetv2/main_tf2.py [hunk 3]
```diff
@@ -117,190 +91,341 @@ def test_step(self, data):
...
+    instance.
+    """
+    logging.info('Setting up training regime for efficientnetv2')
+
+    self.train_split = self.config.train.split or 'train'
+    self.eval_split = self.config.eval.split or 'eval'
+    self.num_train_images = self.config.data.splits[self.train_split].num_images
+    self.num_eval_images = self.config.data.splits[self.eval_split].num_images
+    self._steps_per_epoch = self.num_train_images // self.config.train.batch_size
+    self._total_steps = self._steps_per_epoch * self.config.train.epochs
```
Comment (kartik4949) [inline efficientnetv2/main_tf2.py:189]: @mingxingtan 
total_steps = max(config.train.min_steps, steps_per_epoch*config.train.epochs)
as in tf1 script to use min_steps

File: efficientnetv2/smoke_test_tf2.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  @classmethod
+  def setUpClass(cls):
+    super().setUpClass()
+    FLAGS.tpu = ''
+    FLAGS.model_dir = tempfile.mkdtemp()
+    FLAGS.data_dir = 'null'
+    cls.hparam_str = (
+        'train.batch_size=2,eval.batch_size=2,train.epochs=1,'
+        'train.stages=0,train.lr_base=0,data.splits.train.num_images=2,data.splits.eval.num_images=6,')
```
Comment (kartik4949) [inline efficientnetv2/smoke_test_tf2.py:42]: @mingxingtan a hack to get steps_per_epoch = 1

File: efficientnetv2/smoke_test_tf2.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    tf.compat.v1.reset_default_graph()
+    FLAGS.mode = 'eval'
+    main_lib.main(None)
+ 
+  # TODO(someone): remove `runtime.strategy` and use `use_tpu`
+  @flagsaver.flagsaver(
+      model_name='efficientnetv2-s', dataset_cfg='ImageNet')
+  def test_cpu_b0_model_single_step(self):
+    self._run_single_step_train_and_eval('runtime.strategy=cpu')
```
Comment (kartik4949) [inline efficientnetv2/smoke_test_tf2.py:58]: tf1 script uses flags.use_tpu and tf2 use runtime.strategy, should be make it same?
@mingxingtan

**Pull request closed without merging**